### PR TITLE
Using pydantic-yaml and validations through pydantic.validators.

### PR DESCRIPTION
**Pull request by bsieber-mozilla**
This PR address issue #20

### CODE REVIEW

File: src/templates/powered_by_template.html [hunk 1]
```diff
@@ -20,7 +20,7 @@ <h1>{{title}}</h1>
                 </thead>
                 <tbody class="list">
                     {% for key, value in data.items() %}
-                        {% if enable_query is none or value.enabled is enable_query %}
+                        {% if enable_query == none or value.enabled == enable_query %}
```
Comment (leplatrem) [inline src/templates/powered_by_template.html:23]: Why this change?

File: src/jbi/configuration.py [hunk 2]
```diff
@@ -1,83 +1,33 @@
...
-    """Errors used to determine when the Configuration is invalid"""
-
-
-class ProcessError(Exception):
-    """Error when an exception occurs during processing"""
+    """Error when an exception occurs during processing config"""
 
 
+@lru_cache
 def get_yaml_configurations(
```
Comment (leplatrem) [inline src/jbi/configuration.py:21]: I think it would be clearer if we would use the term *actions configuration* throughout the whole code base, instead of *yaml configuration*.

For example, this function could be called `get_actions()`, especially since it's in the configuration module. 

`actions = configuration.get_actions()`

File: src/jbi/model.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+from types import ModuleType
+from typing import Any, Dict, Optional
+
+from pydantic import Extra, ValidationError, root_validator, validator
+from pydantic_yaml import YamlModel
+
+
+class ActionConfig(YamlModel, extra=Extra.allow):
+    """
+    ActionConfig is the inner model of `actions` in yaml
```
Comment (leplatrem) [inline src/jbi/model.py:15]: This comment is a bit too similar to the one of `Actions`.

```suggestion
    ActionConfig is the inner model for each action in the configuration file
```

File: src/jbi/model.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        except ImportError as exception:
+            raise ValidationError(f"unknown action `{action}`.") from exception
+        except (TypeError, AttributeError) as exception:
+            raise ValidationError("action is not properly setup.") from exception
+        return values
+
+
+class Actions(YamlModel):
+    """
+    Actions is the overall model from parsing the yaml config `actions` sections
```
Comment (leplatrem) [inline src/jbi/model.py:46]:
```suggestion
    Actions is the overall model for the list of `actions` in the configuration file
```

File: src/jbi/model.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import importlib
+from inspect import signature
+from types import ModuleType
+from typing import Any, Dict, Optional
+
+from pydantic import Extra, ValidationError, root_validator, validator
+from pydantic_yaml import YamlModel
+
+
+class ActionConfig(YamlModel, extra=Extra.allow):
```
Comment (leplatrem) [inline src/jbi/model.py:13]: Shouldn't it be just `Action` (since we have `Actions` below) ?

File: src/jbi/model.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+class Actions(YamlModel):
+    """
+    Actions is the overall model from parsing the yaml config `actions` sections
+    """
+
+    actions: Dict[str, ActionConfig]
+
+    @validator("actions")
+    def validate_action_yaml_jbi_naming(
```
Comment (leplatrem) [inline src/jbi/model.py:52]:
```suggestion
    def validate_action_matches_whiteboard(
```

File: src/jbi/router.py [hunk 7]
```diff
@@ -36,7 +36,7 @@ def get_whiteboard_tag(
     whiteboard_tag: Optional[str] = None,
 ):
     """API for viewing whiteboard_tags"""
-    data = configuration.get_yaml_configurations()
+    data = configuration.get_yaml_configurations().dict()["actions"]
```
Comment (leplatrem) [inline src/jbi/router.py:39]: Maybe this helper on the configuration module should return the dict directly

File: src/templates/powered_by_template.html (hunk 1 above)
Comment (bsieber-mozilla) [reply src/templates/powered_by_template.html:23]: https://github.com/pallets/jinja/issues/520#issuecomment-164306945
`is` is a keyword--I was running into 500s when using the dev config and modifiying the query param; so went with `==`

File: src/jbi/model.py (hunk 5 above)
Comment (bsieber-mozilla) [reply src/jbi/model.py:13]: That works for me. I think the naming has evolved over the time--and I might rename the config files to action.*.yaml as well

File: src/jbi/router.py (hunk 7 above)
Comment (bsieber-mozilla) [reply src/jbi/router.py:39]: I did that initially but I was wondering whether we want to be using objects or dictionary's to access the underlying data. 

I can migrate this to a helper for the time being.

**Comment by bsieber-mozilla** [comment]
Thanks for your review @leplatrem I made the suggested changes in the most recent commit

File: src/jbi/configuration.py [hunk 8]
```diff
@@ -1,83 +1,38 @@
...
-
-
-def get_yaml_configurations(
+@lru_cache
+def get_actions(
     jbi_config_file: str = f"config/config.{settings.env}.yaml",
-) -> Dict[str, Dict]:
-    """Convert and validate YAML configuration to python dict"""
+) -> Actions:
+    """Convert and validate YAML configuration to Action object"""
```
Comment (leplatrem) [inline src/jbi/configuration.py:24]:
```suggestion
    """Convert and validate YAML configuration to `Action` objects"""
```

File: src/jbi/configuration.py [hunk 9]
```diff
@@ -1,83 +1,38 @@
...
-        if not hasattr(action_module, "init"):
-            raise TypeError("Module is missing `init` method.")
 
-        signature(action_module.init).bind(**action_parameters)  # type: ignore
-    except ImportError as exception:
-        raise ConfigError(f"Unknown action `{action}`.") from exception
-    except (TypeError, AttributeError) as exception:
-        raise ConfigError("Action is not properly setup.") from exception
+def get_actions_dict():
+    """Returns dict of"""
```
Comment (leplatrem) [inline src/jbi/configuration.py:37]:
```suggestion
    """Returns dict of `get_actions()`"""
```

**Review by leplatrem — APPROVED** [review]

**Review by leplatrem — APPROVED** [review]
